src/systems/systems_wasm/src/aspects.rs

```diff
@@ -0,0 +1,852 @@
+use crate::{
+    f64_to_sim_connect_32k_pos, sim_connect_32k_pos_to_f64, Aspect, MsfsVariableRegistry, Variable,
+};
+use enum_dispatch::enum_dispatch;
+use msfs::sim_connect::{SimConnect, SimConnectRecv, SIMCONNECT_OBJECT_ID_USER};
+use std::error::Error;
+use std::time::{Duration, Instant};
+use systems::simulation::VariableIdentifier;
+
+/// Type used to configure and build an [Aspect].
+///
+/// It should be noted that the resulting [Aspect] executes its tasks in the order in which they
+/// were declared. Declaration order is important when one action depends on another action.
+/// When e.g. a variable that is mapped should then be written to an event, be sure to
+/// declare the mapping before the event writing.
+pub struct MsfsAspectBuilder<'a, 'b> {
+    sim_connect: &'a mut SimConnect<'b>,
+    variables: &'a mut MsfsVariableRegistry,
+    message_handlers: Vec<MessageHandler>,
+    actions: Vec<(VariableAction, ExecuteOn)>,
+}
+
+impl<'a, 'b> MsfsAspectBuilder<'a, 'b> {
+    pub fn new(
+        sim_connect: &'a mut SimConnect<'b>,
+        variables: &'a mut MsfsVariableRegistry,
+    ) -> Self {
+        Self {
+            sim_connect,
+            variables,
+            message_handlers: Default::default(),
+            actions: Default::default(),
+        }
+    }
+
+    pub fn build(self) -> MsfsAspect {
+        let aspect = MsfsAspect::new(self.message_handlers, self.actions);
+
+        aspect
+    }
+
+    /// Initialise the variable with the given value.
+    pub fn init_variable(&mut self, variable: Variable, value: f64) {
+        Self::precondition_not_aircraft_variable(&variable);
+
+        let identifier = self.variables.register(&variable);
+        self.variables.write(&identifier, value);
+    }
+
+    /// Copy a variable's value to another variable.
+    pub fn copy(&mut self, input: Variable, output: Variable) {
+        Self::precondition_not_aircraft_variable(&output);
+
+        let input = self.variables.register(&input);
+        let output = self.variables.register(&output);
+
+        self.actions.push((
+            Map::new(input, |value| value, output).into(),
+            ExecuteOn::PreTick,
+        ));
+    }
+
+    /// Map a variable's value to another variable, applying the given function in the process.
+    pub fn map(
+        &mut self,
+        execute_on: ExecuteOn,
+        input: Variable,
+        func: fn(f64) -> f64,
+        output: Variable,
+    ) {
+        Self::precondition_not_aircraft_variable(&output);
+
+        let inputs = self.variables.register(&input);
+        let output = self.variables.register(&output);
+
+        self.actions
+            .push((Map::new(inputs, func, output).into(), execute_on));
+    }
+
+    /// Map a set of variable values to another variable.
+    pub fn map_many(
+        &mut self,
+        execute_on: ExecuteOn,
+        inputs: Vec<Variable>,
+        func: fn(&[f64]) -> f64,
```

Comment:
I assume it is not trivial to use const generics here to ensure that `func` takes an array with the same number of elements as `inputs` has? Would be nice though. 😄

Reply:
It is not possible. The reason for that is that it would require all usages of `map_many` to have the same number of input variables known at compile time, which is unlikely to be the case.